static_frame/core/frame.py

```diff
@@ -5534,6 +5628,103 @@ def to_parquet(self,
+                elif isinstance(obj, Index):
+                    return {b'sf':clsname,
+                            b'name':obj.name,
+                            b'data':packb(obj.values)} #recurse packb
+                elif isinstance(obj, TypeBlocks):
+                    return {b'sf':clsname,
+                            b'blocks':packb(obj._blocks)} #recurse packb
+            elif package == 'numpy':
+                #msgpack_numpy is breaking with these data types, overriding here
+                if isinstance(obj, np.ndarray):
+                    #Hypothesis coverage:
+                
+                    if obj.dtype.type == np.object_:
+                        typeset = list(set(map(type, obj)))
+                        if len(typeset) > 1:
+                            data = []
+                            for a in obj:
+                                c = a.__class__.__name__
+                                p = a.__class__.__module__.split('.',1)[0]
+                                if p == 'datetime':
+                                    d = a.strftime('%Y %a %b %d %H:%M:%S:%f')
+                                    year = d.split(' ',1)[0].zfill(4) #datetime returns inconsistent year string for <4 digit years on some systems
+                                    d = year+' '+d.split(' ',1)[-1]
+                                elif c == 'bool':
+                                    d = int(a)
+                                else:
+                                    d = str(a)
+                                data.append((p+'.'+c, d))
+                            return {b'np': True,
+                                    b'dtype': 'multitype',
+                                    b'data': packb(data)} #recurse packb
+                        else:
+                            t = typeset[0]
+                            tname = t.__name__
+                            if tname in ['datetime', 'date']:
+                                data = []
+                                for a in obj:
+                                    d = a.strftime('%Y %a %b %d %H:%M:%S:%f')
+                                    year = d.split(' ',1)[0].zfill(4) #datetime returns inconsistent year string for <4 digit years on some systems
+                                    d = year+' '+d.split(' ',1)[-1]                              
+                                    data.append(d)
+                            elif tname == 'int': #msgpack fails on a large number case from hypothesis here
+                                data = [str(a) for a in obj]
+                            elif tname == 'Fraction':
+                                data = [str(a) for a in obj.astype(t)]
+                            else:
+                                data = [a for a in obj]
+                            return {b'np': True,
+                                    b'dtype': tname,
+                                    b'data': packb(data)} #recurse packb
```

Comment:
We can assume arrays with object dtype always have hetergenous types, and thus only keep the branch where `len(typeset) > 1`: that we do not have to duplicate the handling for all types here.

But even then it seems like we are doing a lot of extra work here for object arrays. Can we simply use `array.tolist()` and let msgpack handle the list as it would for any other list of objects? This might be tricky, on the decode side, to make sure we re-create the array.

Reply:
I refactored the code to remove the excess logic branch, and I've moved the fix into its own function. I couldn't come up with a way to effectively use `arra.tolist()` to avoid the extra code, as hypothesis includes tests with nested ndarrays. However I have minimized the fix as much as possible.
[https://github.com/InvestmentSystems/static-frame/blob/795143b92a8de9fec749a1f90dd48ef7d504f789/static_frame/core/frame.py#L5626-L5647](https://github.com/InvestmentSystems/static-frame/blob/795143b92a8de9fec749a1f90dd48ef7d504f789/static_frame/core/frame.py#L5626-L5647)

I didn't want to give the same treatment to the static_frame objects, because then the convenience of simply calling up the class objects from `globals()[clsname]` would be destroyed, and that would create extra maintenance for any new sub-classes added to static-frame.

Reply:
Sure, this is good for now. We might be able to revisit this if `msgpack_numpy` improves support for object arrays in the future.